sources/api_client.py

```diff
@@ -0,0 +1,217 @@
+        self.next_reference = response.links.get(self.links_next_key, {}).get("url")
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json()
+
+
+class JSONResponsePaginator(BaseNextUrlPaginator):
+    """A paginator that uses a specific key in the JSON response to find
+    the next page URL.
+    """
+
+    def __init__(self, next_key: str = "next", records_key: str = "results"):
+        """
+        Args:
+            next_key (str, optional): The key in the JSON response that
+                contains the next page URL. Defaults to 'next'.
+            records_key (str, optional): The key in the JSON response that
+                contains the page's records. Defaults to 'results'.
+        """
+        super().__init__()
+        self.next_key = next_key
+        self.records_key = records_key
+
+    def update_state(self, response: Response):
+        self.next_reference = response.json().get(self.next_key)
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json().get(self.records_key, [])
+
+
+class BearerTokenAuth(AuthBase):
+    def __init__(self, token: str):
+        self.token = token
+
+    def __call__(self, request):
+        request.headers["Authorization"] = f"Bearer {self.token}"
+        return request
+
+
+def join_url(base_url: str, path: str) -> str:
+    if not base_url.endswith("/"):
+        base_url += "/"
+    return base_url + path.lstrip("/")
+
+
+class APIClient:
+    """A generic API client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthBase] = None,
+        paginator: Optional[BasePaginator] = None,
+    ) -> None:
+        self.base_url = base_url
+        self.headers = headers
+        self.auth = auth
+        self.paginator = paginator if paginator else HeaderLinkPaginator()
+
+    def make_request(self, path="", method="get", params=None, json=None):
+        if path.startswith("http"):
+            url = path
+        else:
+            url = join_url(self.base_url, path)
+
+        response = requests.request(
+            method=method,
+            url=url,
+            headers=self.headers,
+            params=params if method.lower() == "get" else None,
+            json=json if method.lower() in ["post", "put"] else None,
+            auth=self.auth,
+        )
+        response.raise_for_status()
```

Comment:
should we also allow not to fail for example cases when users want to get successful results and process erred results later?

Reply:
This will probably require some sort of monadic type `Result`